Comments from a read-through

I went through this lesson as I will work through the material that comes later and I wanted to get an idea of what the participants have gone through. It is really very nice. I understand that it is still in development but I'm however listing here some minor remarks that I made when working through the lesson, feel free to take them into account or not 😃 

- [x] The entry-page introduction starts with "You’ve have recently learned about CI/CD through GitHub and Travis CI". This is a probably a leftover from an earlier instance of this tutorial. This time it does not look to me that there is anything like that in the pre-workshop material

In 02-enter-sandman:
- [x] before the first python example, although obvious probably for almost all, you could add a short note that it is enough to type python on command line to get to that environment (and exit() to exit...) 
- [x] the example of setupATLAS gives me (as a CMS user)
```
bash-4.2$ setupATLAS || echo ignore setupATLAS
bash: setupATLAS: command not found
ignore setupATLAS
```
(whether that matters)

in 04-understanding-yaml:
- [x] Maybe, after "YAML is Yet Another Markup Language that is usually the modus operandi for CI systems." add a simple phrase that describes what it does in context of CI to motivate the reader to read through and also mention that they are use elsewhere as well
- [x] typo in "howver two spaces "
- [x] It would be useful to have a brief explanation of the scope of the listing (in YAML) below Scalars. Do I get it correctly that it is there to illustrate the syntax of different things in YAML? It would maybe be good to start with a typical example YAML file in the CI context.
- [x] the anchor stuff requires a bit of brain gymnastics (which is probably not a bad thing)

in 05-understanding-yaml-and-ci
- [x] aha 💡  , now comes the YAML file in the CI context which I was expecting. For my learning type, I would have preferred to start with this, but I understand that other people may prefer to have the generic YAML introduction before
- [x] analysisbase is not known to CMS people but it probably does not matter here
- [x] at the end: "The API for all GitLab CI/CD pipeline configurations" -> "The reference guide for all GitLab CI/CD pipeline configurations"?

07-hello-world-ci
- [x] for easier copy-pasting, please remove > at the beginnning of the lines
- [x] this is the first time I hear of "linter", maybe a brief explanation?
- [ ] I must be missing something, I pushed the (empty-as-instructed) file into the repo but I do not see what is in the instructions but this:
![aw-add-ci](https://user-images.githubusercontent.com/5859918/74435102-feb19080-4e63-11ea-9fd4-f81d126acc16.JPG)
- [x] it looks as the CI does not run if the file is empty
- [x] Indeed, if I add the hello world example, it runs nicely and clicking on CI/CD Pipelines I see this:
![aw-pipelines](https://user-images.githubusercontent.com/5859918/74437037-beeca800-4e67-11ea-860f-a9ba65d9c812.JPG)

More may follow...

in 08-adding-ci-to-your-code:
- [x] having done https://awesome-workshop.github.io/awesome-htautau-analysis/07-preparation/index.html and in my local `awesome-analysis-eventselection` on lxplus:
```
bash-4.2$ COMPILER=$(root-config --cxx)
bash-4.2$ $COMPILER -g -O3 -Wall -Wextra -Wpedantic -o skim skim.cxx
skim.cxx:11:31: fatal error: ROOT/RDataFrame.hxx: No such file or directory
 #include "ROOT/RDataFrame.hxx"
                               ^
compilation terminated.
```
This probably needs to be checked against https://awesome-workshop.github.io/awesome-htautau-analysis/03-skimming/index.html

[UPDATE:]
- [ ] OK, reading forward, now I got it. However, I did instinctively test on local machine before starting to adding things to .`gitlab-ci.yml` and got stuck here because the compilation did not work.
Maybe add that this is expected to reduce the confusion...
- [ ] I would suggest to bring the compilation flag issue up to the beginning and start working on the .`gitlab-ci.yml` file only when the local compilation works. That would reflect the normal workflow.

--- a/_episodes/01-introduction.md
+++ b/_episodes/01-introduction.md
@@ -46,8 +46,8 @@ Don't just limit yourself to thinking of CI/CD as primarily for testing changes,
 
 Now, obviously, we're not going to make our own fully-fledged CI/CD solution. Plenty exist in the wild today, and below are just a popular few:
 
-- [Native GitLab CI/CD](https://about.gitlab.com/)
-- (beta) [Native GitHub CI/CD](https://github.blog/2019-08-08-github-actions-now-supports-ci-cd/)
+- [Native GitLab CI/CD](https://docs.gitlab.com/ee/ci/)
+- [Native GitHub CI/CD](https://github.com/features/actions)
 - [Travis CI](https://travis-ci.org/)
 - [Circle CI](https://circleci.com/)
 - [TeamCity](https://www.jetbrains.com/teamcity/)

--- a/_episodes/02-enter-sandman.md
+++ b/_episodes/02-enter-sandman.md
@@ -152,10 +152,10 @@ and then make it executable `chmod +x python_exit.py`. Now, try running it with 
 
 # Ignoring Exit Codes
 
-To finish up this section, one thing you'll notice sometimes in ATLAS is that a script you run doesn't seem to respect exit codes. A notable example is the use of `setupATLAS` which returns non-zero exit status codes even though it runs successfully! This can be very annoying when you start development with the assumption that exit status codes are meaningful (such as with CI). In these cases, you'll need to ignore the exit code. An easy way to do this is to execute a second command that always gives `exit 0` if the first command doesn't, like so:
+To finish up this section, one thing you'll notice sometimes (in ATLAS or CMS) is that a script you run doesn't seem to respect exit codes. A notable example in ATLAS is the use of `setupATLAS` which returns non-zero exit status codes even though it runs successfully! This can be very annoying when you start development with the assumption that exit status codes are meaningful (such as with CI). In these cases, you'll need to ignore the exit code. An easy way to do this is to execute a second command that always gives `exit 0` if the first command doesn't, like so:
 
 ~~~
-> setupATLAS || echo ignore setupATLAS
+> :(){ return 1; };: || echo ignore failure
 ~~~
 {: .source}
 

--- a/_episodes/02-enter-sandman.md
+++ b/_episodes/02-enter-sandman.md
@@ -91,6 +91,13 @@ Try out some other commands on your system, and see what things look like.
 
 As you've seen above, the exit code from the last executed command is stored in the `$?` environment variable. Accessing from a shell is easy `echo $?`. What about from python? There are many different ways depending on which library you use. Using similar examples above, we can use the (note: deprecated) `os.system` call:
 
+> ## Snake Charming
+>
+> To enter the Python interpreter, simply type `python` in your command line.
+>
+> Once inside the Python interpreter, simply type `exit()` then press enter, to exit.
+{: .callout}
+
 ~~~
 >>> import os,subprocess
 >>> ret = os.system('ls')

--- a/_episodes/03-understanding-yaml.md
+++ b/_episodes/03-understanding-yaml.md
@@ -19,7 +19,7 @@ YAML is Yet Another Markup Language is a human-readable data-serialization langu
 > ## Tabs or Spaces?
 >
 > We strongly suggest you use spaces for a YAML document. Indentation is done
-> with one or more spaces, howver **two spaces** is the unofficial standard
+> with one or more spaces, however **two spaces** is the unofficial standard
 > commonly used.
 {: .callout}
 

--- a/_episodes/03-understanding-yaml.md
+++ b/_episodes/03-understanding-yaml.md
@@ -14,7 +14,7 @@ keypoints:
 
 # YAML
 
-YAML is Yet Another Markup Language that is usually the modus operandi for CI systems. We'll cover, briefly, some of the native types involved and what the structure looks like.
+YAML is Yet Another Markup Language is a human-readable data-serialization language. It is commonly used for configuration files and in applications where data is being stored or transmitted. CI systems' modus operandi typically rely on YAML for configuration. We'll cover, briefly, some of the native types involved and what the structure looks like.
 
 > ## Tabs or Spaces?
 >

--- a/_episodes/04-understanding-yaml-and-ci.md
+++ b/_episodes/04-understanding-yaml-and-ci.md
@@ -128,7 +128,7 @@ job two:
 
 ## Reference
 
-The API for all GitLab CI/CD pipeline configurations is found at [https://docs.gitlab.com/ee/ci/yaml/](https://docs.gitlab.com/ee/ci/yaml/). This contains all the different parameters you can assign to a job.
+The reference guide for all GitLab CI/CD pipeline configurations is found at [https://docs.gitlab.com/ee/ci/yaml/](https://docs.gitlab.com/ee/ci/yaml/). This contains all the different parameters you can assign to a job.
 
 > ## Further Reading
 > - [https://docs.gitlab.com/ee/ci/yaml/](https://docs.gitlab.com/ee/ci/yaml/)

--- a/_episodes/04-understanding-yaml-and-ci.md
+++ b/_episodes/04-understanding-yaml-and-ci.md
@@ -19,18 +19,16 @@ keypoints:
 The GitLab CI configurations are specified using a YAML file called `.gitlab-ci.yml`. Here is an example:
 
 ~~~
-image: atlas/analysisbase:latest
+image: rikorose/gcc-cmake
 
 before_script:
-  - source /home/atlas/release_setup.sh
   - mkdir build
 
-build_analysis:
+build_code:
   script:
     - cd build
     - cmake ../src
     - cmake --build .
-    - source "${AnalysisBase_PLATFORM}/setup.sh"
 ~~~
 {: .language-yaml}
 
@@ -39,12 +37,12 @@ build_analysis:
 > Sometimes, `script` commands will need to be wrapped in single or double quotes. For example, commands that contain a colon (`:`) need to be wrapped in quotes so that the YAML parser knows to interpret the whole thing as a string rather than a “key: value” pair. Be careful when using special characters: `:`, `{`, `}`, `[`, `]`, `,`, `&`, `*`, `#`, `?`, `|`, `-`, `<`, `>`, `=`, `!`, `%`, `@`, `\``.
 {: .callout}
 
-This is the simplest possible configuration that will work for most analysis codes using `AnalysisBase` on the latest release:
+This is the simplest possible configuration that will work for most code using minimal dependencies with `cmake` and `make`:
 
-1. Define one job `build_analysis` (the job names are arbitrary) with different commands to be executed.
+1. Define one job `build_code` (the job names are arbitrary) with different commands to be executed.
 2. Before every job, the commands defined by `before_script` are executed.
 
-The `.gitlab-ci.yml` file defines sets of jobs with constraints of how and when they should be run. The jobs are defined as top-level elements with a name (in our case `build_analysis`) and always have to contain the `script` keyword. Let's explore this structure in more depth.
+The `.gitlab-ci.yml` file defines sets of jobs with constraints of how and when they should be run. The jobs are defined as top-level elements with a name (in our case `build_code`) and always have to contain the `script` keyword. Let's explore this structure in more depth.
 
 ## Overall Structure
 
@@ -86,7 +84,7 @@ What can you not use as job names? There are a few reserved keywords (because th
 Global parameters mean that you can set parameters at the top-level of the YAML file. What does that actually mean? Here's another example:
 
 ~~~
-image: atlas/analysisbase:21.2.62
+image: rikorose/gcc-cmake
 
 stages: [build, test, deploy]
 
@@ -94,7 +92,7 @@ job1:
   script: make
 
 job2:
-  image: atlas/analysisbase:21.2.72
+  image: rikorose/gcc-cmake:gcc-6
   script: make
 ~~~
 {: .language-yaml}
@@ -122,7 +120,7 @@ job two:
 > ~~~
 > job:
 >   image:
->     name: atlas/analysisbase:21.2.62
+>     name: rikorose/gcc-cmake:gcc-6
 > ~~~
 > {: .language-yaml}
 > where the colon refers to a child key.

--- a/_episodes/04-understanding-yaml-and-ci.md
+++ b/_episodes/04-understanding-yaml-and-ci.md
@@ -116,7 +116,7 @@ job two:
 
 > ## Understanding the Reference
 >
-> One will notice that the reference uses colons like `image:name` to refer to parameter names. This is represented in yaml like:
+> One will notice that the reference uses colons like `:job:image:name` to refer to parameter names. This is represented in yaml like:
 > ~~~
 > job:
 >   image:

--- a/_episodes/06-hello-world-ci.md
+++ b/_episodes/06-hello-world-ci.md
@@ -55,6 +55,11 @@ We should fix this. If you click through again on the red `x` on the left for th
 > This can also be found by going to `CI/CD -> Pipelines` or `CI/CD -> Jobs` page and clicking the `CI Lint` button at the top right.
 {: .callout}
 
+> ### But what's a linter?
+>
+> From [wikipedia](https://en.wikipedia.org/wiki/Lint_(software)): `lint`, or a linter, is a tool that analyzes source code to flag programming errors, bugs, stylistic errors, and suspicious constructs. The term originates from a Unix utility that examined C language source code.
+{: .callout}
+
 Lastly, we'll open up a merge request for this branch, since we plan to merge this back into master when we're happy with the first iteration of the CI/CD.
 
 > ## Work In Progress?

--- a/_episodes/06-hello-world-ci.md
+++ b/_episodes/06-hello-world-ci.md
@@ -19,7 +19,7 @@ We've been working on the analysis code which has a lot of work done, but we sho
 
 ~~~
 cd awesome-analysis-eventselection/
-touch .gitlab-ci.yml
+echo "hello world" >> .gitlab-ci.yml
 git checkout -b feature/add-ci
 git add .gitlab-ci.yml
 git commit -m "my first ci/cd"
@@ -72,7 +72,7 @@ Lastly, we'll open up a merge request for this branch, since we plan to merge th
 
 ## Fixing the YAML
 
-Now, our YAML is currently invalid, but this makes sense because we didn't actually define any jobs to run. Let's go ahead and create our first job that simply echoes "Hello World".
+Now, our YAML is currently invalid, but this makes sense because we didn't actually define any script to run. Let's go ahead and update our first job that simply echoes "Hello World".
 
 ~~~
 hello world:

--- a/_episodes/06-hello-world-ci.md
+++ b/_episodes/06-hello-world-ci.md
@@ -18,12 +18,12 @@ keypoints:
 We've been working on the analysis code which has a lot of work done, but we should be good physicists (and people) by adding tests and CI/CD. The first thing we'll do is create a `.gitlab-ci.yml` file in the project. You should have two projects, one with the analysis code and one for the statistical fits, which I'll call `awesome-analysis-eventselection` and `awesome-analysis-statistics`.
 
 ~~~
-> cd awesome-analysis-eventselection/
-> touch .gitlab-ci.yml
-> git checkout -b feature/add-ci
-> git add .gitlab-ci.yml
-> git commit -m "my first ci/cd"
-> git push -u origin feature/add-ci
+cd awesome-analysis-eventselection/
+touch .gitlab-ci.yml
+git checkout -b feature/add-ci
+git add .gitlab-ci.yml
+git commit -m "my first ci/cd"
+git push -u origin feature/add-ci
 ~~~
 {: .source}
 

--- a/_episodes/10-the-worlds-a-stage.md
+++ b/_episodes/10-the-worlds-a-stage.md
@@ -41,6 +41,19 @@ build_skim_latest:
 
 We're going to talk about another global parameter `:stages` (and the associated per-job parameter `:job:stage`. Stages allow us to group up parallel jobs with each group running after the other in the order you define. What have our jobs looked like so far in the pipelines we've been running?
 
+> ## Understanding the Reference
+>
+> One will notice that the reference uses colons like `:job:image:name` to refer to parameter names. This is represented in yaml like:
+> ~~~
+> job:
+>   image:
+>     name: rikorose/gcc-cmake:gcc-6
+> ~~~
+> {: .language-yaml}
+> where the colon refers to a child key.
+{: .callout}
+
+
 ![CI/CD Default Stages in Pipeline]({{site.baseurl}}/fig/ci-cd-default-stages.png)
 
 > ## Default Stage

--- a/_episodes/12-the-spy-game.md
+++ b/_episodes/12-the-spy-game.md
@@ -85,7 +85,7 @@ Let's go ahead and add some custom variables to fix up our access control.
 
 # Adding `kinit` for access control
 
-Now it's time to update your CI/CD to use the environment variables you defined by adding `echo $SERVICE_PASS | kinit $CERN_USER@CERN.CH` as part of the `before_script` to the `skim_ggH` job as that's the job that requires access.
+Now it's time to update your CI/CD to use the environment variables you defined by adding `printf $SERVICE_PASS | base64 -d | kinit $CERN_USER@CERN.CH` as part of the `before_script` to the `skim_ggH` job as that's the job that requires access.
 
 # Adding Artifacts on Success
 

--- a/_episodes/12-the-spy-game.md
+++ b/_episodes/12-the-spy-game.md
@@ -37,6 +37,15 @@ ERROR: Job failed: exit code 1
 
 So we need to give our CI/CD access to our data. This is actually a good thing. It means CMS can't just grab it! Anyhow, this is done by pretty much done by executing `printf $SERVICE_PASS | base64 -d | kinit $CERN_USER` assuming that we've set the corresponding environment variables by safely encoding them (`printf "hunter42" | base64`).
 
+> ## Running examples with variables
+>
+> Sometimes you'll run into a code example here that you might want to run locally but relies on variables you might not have set? Sure, simply do the following
+> ~~~
+> SERVICE_PASS=hunter42 CERN_USER=GoodWill printf $SERVICE_PASS | base64 -d | kinit $CERN_USER
+> ~~~
+> {: .language-bash}
+{: .callout}
+
 > ## Base-64 encoding?
 >
 > Sometimes you have a string that contains certain characters that would be interpreted incorreectly by GitLab's CI system. In order to protect against that, you can safely base-64 encode the string, store it, and then decode it as part of the CI job. This is entirely safe and recommended.

--- a/index.md
+++ b/index.md
@@ -19,7 +19,7 @@ permalink: index.html  # Is the only page that doesn't follow the pattern /:path
 Introduction
 ------------
 
-You've have recently learned about CI/CD through GitHub and Travis CI. In ATLAS we use uses GitLab to host our code. GitLab is bundled with a built-in CI/CD system that we'll learn how to develop on to make our code robust to errors, preserved, and reproducible.
+At CERN, we use GitLab to host our code. GitLab is bundled with a built-in CI/CD system that we'll learn how to develop on to make our code robust to errors, preserved, and reproducible.
 
 The aim of this module is to:
 - explore what it means to build a CI/CD workflow

